### ISSUE TITLE
Swift 3 Release?

### ISSUE THREAD

**Issue by rlam3** [SwiftyJSON/SwiftyJSON#627]
Swift 3 Release?

**Comment by appsailor** [comment]
Try the swift3 branch from my fork.

**Comment by xudafeng** [comment]
💯

**Comment by Dark0rion** [comment]
update to swift 3 please!

**Comment by appsailor** [comment]
@Dark0rion, If you use the following in Podfile, you can survive till the master is ready.
pod 'SwiftyJSON', :git => 'https://github.com/appsailor/SwiftyJSON.git', :branch => 'swift3'

**Comment by flandy84** [comment]
hi @appsailor i just tried your swift3-branch on the new xcode8-(final/appstore release).
It builds fine, but there seems to be an issue, when using swifty-json in for-loops and/or subscripts?!

this line works fine in xcode7 (swiftyjson for swift 2.2) but ends up an a endless loop (growing memory-footprint with empty strings).

        for (_, subJson): (String, JSON) in json["someMissingKeyInJSONDictionary"] {
            let infoString = subJson.stringValue
            infotext.append(infoString)
        }

hmm, any idea?

**Comment by appsailor** [comment]
@flandy84, it would be nice to get a sample json file that can be used as a test file. Would you be able to share it so that I can take a look? Thx -aldrin

**Comment by flandy84** [comment]
hi @appsailor. Sorry, I'm on a journey the next days, but my developer explained me, you can use any json-dictionary to reproduce this issue; a arbitury JSON won't have the key "someMissingKeyInJSONDictionary" with an expected Array-value in it ;)

**Comment by ossmalpha** [comment]
For me nor the master, nor swift3 branches are working with Swift 3 on Xcode 8. But I guess that's no surprise because last push was around 3 months ago. When is this going to be updated to work on new version?

**Comment by kylebrowning** [comment]
@ossmalpha We simply do not know. Owners are MIA.

**Comment by pflanny2** [comment]
@ossmalpha is it building for you? we have just started a new project in 3 and we can get it to build with the Swift 3 branch, but I don't know if we are implementing it anywhere yet. It would be helpful to know if it builds but doesn't work for you (if that makes sense).

**Comment by ossmalpha** [comment]
@pflanny2 No, it's not building for me. I also did create a new project just to check which pods are working and which aren't. For SwiftyJSON I tested both branches and there seemed to be some errors while building in both cases. 

How did you managed to get swift3 branch to build? Just pulled from there and built or converted to Swift 3 as well?

**Comment by pflanny2** [comment]
@ossmalpha 

First, I have the following line in the main section of the podfile:

pod 'SwiftyJSON', :git => 'https://github.com/appsailor/SwiftyJSON.git', :branch => 'swift3'

Then, after 'end' of the main section, I have included the following:

# https://github.com/Alamofire/Alamofire/issues/1526
post_install do |installer|
    installer.pods_project.targets.each do |target|
        target.build_configurations.each do |config|
            config.build_settings['SWIFT_VERSION'] = '3.0'
        end
    end
end

Honestly, I'm not sure if you need the latter part to get SwiftyJSON to work; as you can tell, I got that from an issue with getting AlamoFire to work -- and so perhaps fixed compilation of both pods.

If that doesn't work, you may also want to check the Swift Compiler - Version settings. I have the option "Use Legacy Swift Language Version" to No.

Let me know if that helps.

**Comment by ossmalpha** [comment]
@pflanny2 I had both of the things that you mentioned already written. Also, I have Alamofire, Socket.io and many more pods working with Swift 3 already, so I'm sure that all specs have been set right. 

I guess we need to see whether this is working for others and wait until the code is going to be merged into master.

**Comment by ossmalpha** [comment]
@pflanny2 Actually, it's my bad. Instead of `https://github.com/appsailor/SwiftyJSON.git` I had `https://github.com/SwiftyJSON/SwiftyJSON.git`. So I guess that I shouldn't have copied the clone url.

**Comment by pflanny2** [comment]
@ossmalpha that's interesting; I hadn't noticed that. But when I switched from appsailor's fork to SwiftyJson I started getting a lot of compilation errors.

So does it seem like appsailor's fork is working? Not sure how support will be for either

**Comment by siberianisaev** [comment]
@pflanny2 Xcode still asking for conversion to latest Swift.

`pod 'SwiftyJSON', :git => 'https://github.com/appsailor/SwiftyJSON.git', :branch => 'swift3'` with `config.build_settings['SWIFT_VERSION'] = '3.0'`

Conversion process:
![1](https://dl.dropboxusercontent.com/u/25357310/SwiftyJSON_1.png)
![2](https://dl.dropboxusercontent.com/u/25357310/SwiftyJSON_2.png)
![3](https://dl.dropboxusercontent.com/u/25357310/SwiftyJSON_3.png)

**Comment by pflanny2** [comment]
@siberianisaev, not sure why it's wanting to convert those. Here's the process I did...in case it's helpful at all.

1. Convert everything to Swift 3.
2. Update the pod file (as mentioned in a previous post).
3. Run Pod Update.

After step 3, it didn't ask me to convert anything, and could build/run.

That's my initial response. I'll try to look into things on my end a little more.

**Comment by appsailor** [comment]
@pflanny2, I have fixed those changes as well and pushed those over to my fork.

**Comment by appsailor** [comment]
@flandy84, could you try with the latest from my fork and see it works?

 I have seen the unit tests work okay as per the travis CI logs:
Test Case '-[SwiftyJSON_tvOS_Tests.SubscriptTests testOutOfBounds]' started.
Test Case '-[SwiftyJSON_tvOS_Tests.SubscriptTests testOutOfBounds]' passed (0.001 seconds).
Test Suite 'SubscriptTests' passed at 2016-09-15 20:15:25.976.
	 Executed 16 tests, with 0 failures (0 unexpected) in 0.051 (0.105) seconds
Test Suite 'SwiftyJSON tvOS Tests.xctest' passed at 2016-09-15 20:15:25.977.
	 Executed 111 tests, with 0 failures (0 unexpected) in 3.755 (4.099) seconds
Test Suite 'All tests' passed at 2016-09-15 20:15:25.989.
	 Executed 111 tests, with 0 failures (0 unexpected) in 3.755 (4.115) seconds
** TEST SUCCEEDED **

**Comment by analvikingur** [comment]
@appsailor, I have tested your branch in my project –– it works fine!

**Comment by kocherovets** [comment]
swift 3 branch works very slow with this fragment (few seconds)

    "task_categs": {
        "flag-argentina": ["argentin"],
        "flag-australia": ["australia", "sydney", "melbourne", "perth"],
        "flag-belarus": ["belarus"],
... about 300 items

version for swift 2.2 worked instantly

**Comment by kocherovets** [comment]
https://github.com/IBM-Swift/SwiftyJSON 
version works good for me

**Comment by Gujci** [comment]
Any Carhage users here?
I have tried @appsailor s "swift 3" branch, which would work fine, but  Carthage actually needs a new tag on the brach to be able to check it out. So I have forked it and added a new tag and now it works.
Correct me if my theory was wrong.

**Comment by kylebrowning** [comment]
I got it to work on Carthage will post file when i get home

**Comment by kylebrowning** [comment]
```
git "https://github.com/appsailor/SwiftyJSON.git" "swift3"
github "Alamofire/Alamofire" ~> 4.0
github "radex/SwiftyUserDefaults" ~> 3.0
github "Hearst-DD/ObjectMapper" ~> 2.0
```

**Comment by CouponWallet** [comment]
the swift 3 branch does not compile in Xcode 8 ... do we have an idea when this will be resolved?

**Comment by flandy84** [comment]
@appsailor just tested it. The neverending-for-loop issue still exists in swift3 version of swiftyjson :/

**Comment by cobear25** [comment]
The same thing that's happening to @siberianisaev is happening to me. The swift3 branch still shows things that need converted.

**Comment by piotrros** [comment]
That's why libraries shouldn't be created in swift.. Each time there's a new swift release everything's broken...

**Comment by kylebrowning** [comment]
I believe this is the last code breaking swift up for awhile.

**Comment by analvikingur** [comment]
@piotrros
Even that Objective-C doesn't have deep changes doesn't mean that developers should use it for libraries development

**Comment by DenHeadless** [comment]
Hello, guys.

As all of you, we've been waiting for SwiftyJSON swift3 release, however it became clear, that we don't have any means to reach library authors right now. And since we use not only pod, but have a [framework, that depends on it](https://github.com/MLSDev/TRON), we need a solution.

So, we've forked SwiftyJSON to https://github.com/MLSDev/SwiftyJSON, and are making release ourselves.

We use a new name to not intersect:

```ruby
pod 'SwiftyJSON3', '~> 3.0.0-beta.1'
```

This release is basically all changes from #628. Keep in mind, this is not a permanent, but temporary solution, until authors come back and official release is made.

When that happens, `SwiftyJSON3` pod will immediately be deprecated in favor of original pod, and after transition period will be removed.

**Comment by giacgbj** [comment]
I suggest to try https://github.com/LoganWright/Genome, which seems to be a good alternative to SwiftyJSON.

**Comment by xudafeng** [comment]
@giacgbj 👍

**Comment by mrojas** [comment]
Hi there,
@flandy84 I'm also having the infinite loop. Did you find a way to solve it?

**Comment by henriquewilson1** [comment]
@DenHeadless 👍

**Comment by rodrigoruiz** [comment]
[!] Unable to find a specification for `SwiftyJSON3 (~> 3.0.0-beta.1)`

**Comment by rodrigoruiz** [comment]
I think it works if you add this instead `pod 'SwiftyJSON3', :git => 'https://github.com/MLSDev/SwiftyJSON.git'`, but still has the infinite loop problem

**Comment by DenHeadless** [comment]
@rodrigoruiz Run pod update and you will see it. Also, if you have a fix for infinite loop problem you are talking about, we can fix it in next beta release.

JSON example would be super helpful.

**Comment by scrainie** [comment]
Hey I've had no problems with the swift3 migration from @appsailor ; however just ran into this issue: 

Terminating app due to uncaught exception 'NSInvalidArgumentException', reason: 'Invalid type in JSON write (_SwiftValue)'

Any suggestions, the rest of the calls I'm making work fine!

**Comment by muukii** [comment]
I made to try.
Usage is similar to SwiftyJSON
[https://github.com/muukii/JAYSON](https://github.com/muukii/JAYSON)

**Comment by flandy84** [comment]
@mrojas invinity-loop issue is still in all swift3 forks of swifty json :( have no solution only the ugly workaround posted here (pre for-loop-usage array check )

**Comment by rodrigoruiz** [comment]
So, in my case the "infinite loop issue" (I'm just repeating the name others said, I don't really know what the issue is, only that it's stuck doing something) happens when I call `.sorted(by: { $0.0 < $1.0 })` on a `nil` value

**Comment by rodrigoruiz** [comment]
Now that I see it, it seems weird that it was working before, I'd expect a crash or something.
I'm not sure what was the previous behavior, but my app was working... I'm guessing it was returning an empty array.

**Comment by scrainie** [comment]
Damn! I've finally been hit with the infinite loop ! Odd. Time to find a fix :)

**Comment by scrainie** [comment]
Solved ... check for null ... self.json!["timetable"]["adhoc"].exists() {} bingo..

**Comment by hanzhuzi** [comment]
it is not support Swift3.0?

**Comment by saiwu-bigkoo** [comment]
update to swift 3 please!  update to swift 3 please! update to swift 3 please!
update to swift 3 please!update to swift 3 please!update to swift 3 please!update to swift 3 please!update to swift 3 please!update to swift 3 please!update to swift 3 please!update to swift 3 please!update to swift 3 please!update to swift 3 please!update to swift 3 please!update to swift 3 please!update to swift 3 please!update to swift 3 please!update to swift 3 please!update to swift 3 please!update to swift 3 please!update to swift 3 please!update to swift 3 please!update to swift 3 please!update to swift 3 please!update to swift 3 please!update to swift 3 please!

**Comment by ossmalpha** [comment]
This works for my project perfectly running on Swift 3 in Xcode 8. Use this in your Podfile: 

`pod 'SwiftyJSON3', '~> 3.0.0-beta.1'`

**Comment by Wishercat** [comment]
@DenHeadless  Great job!

**Comment by TakanashiOuken** [comment]
`'SwiftyJSON', :git => 'https://github.com/appsailor/SwiftyJSON.git', :branch => 'swift3'`
`pod 'SwiftyJSON3', '~> 3.0.0-beta.1'`
Both of these don't work. Ask me to Convert

**Comment by scrainie** [comment]
To be honest, I started one of my big projects from scratch . Added each POD separately from 'swift3' branches. Just copied and pasted class files and mine is working. The merge with existing was a nightmare, was actually easier to start fresh. Less tears anyway :)

**Comment by DenHeadless** [comment]
@cobear25 Apparently your Podfile is not configured properly, don't specify git URL or branches, simply use

```ruby
pod 'SwiftyJSON3'
```

and run `pod install`.

Code you are showing on your screenshot is not the code from SwiftyJSON3 repo.

**Comment by mpetar92** [comment]
Any idea why i would be getting the `SwiftyJSON.swift:226:1: Extension of type 'JSON' cannot inherit from class 'Collection'`
on line`extension JSON: Collection` 

I can't find any1 else having this problem, yet on all the branches you guys recommended i get this problem

**Comment by Dbigshooter** [comment]
pod 'SwiftyJSON3', '~> 3.0.0-beta.1' works like a charm, however it, for some odd reason, asks me to convert.

**Comment by TakanashiOuken** [comment]
@Dbigshooter 
After converting it can work?
It still shows me a list of errors.

**Comment by Dbigshooter** [comment]
@DenHeadless It worked like a charm for me yesterday, but when I try to install the pod today I get
Unable to find a specification for `SwiftyJSON3 (~> 3.0.0-beta.1)`?

**Comment by DenHeadless** [comment]
@Dbigshooter podspec is definitely there - https://github.com/CocoaPods/Specs/tree/master/Specs/SwiftyJSON3

Try running `pod repo update` before `pod install`, maybe you were installing on different machine? Starting with CocoaPods 1.0.0 pod install does not trigger pod repo update.

**Comment by Dbigshooter** [comment]
@DenHeadless Now I get:

[!] There are only pre-release versions available satisfying the following requirements:

	'SwiftyJSON3', '>= 0'

You should explicitly specify the version in order to install a pre-release version

**Comment by DenHeadless** [comment]
try following syntax in your podfile: 

```ruby
pod 'SwiftyJSON3', '~> 3.0.0-beta.1'
```

**Comment by Dbigshooter** [comment]
@DenHeadless Worked like a charm, thanks! Do you have a explanation on why I need to set "Use legacy Swift language version" to NO for almost all my libs every time I update?

**Comment by DenHeadless** [comment]
So happy authors are here, and this issue is finally resolved!

As promised, `SwiftyJSON3` pod is now deprecated. Please migrate to `SwiftyJSON` if you've been using it. Thank you all!

**Comment by rodrigoruiz** [comment]
I don't see any releases on the library since October 2015, am I missing something?

**Comment by zhigang1992** [comment]
@rodrigoruiz I've only released the 3.0.0 tag and pointing cocoapods to it. Will add release later today, still trying to clean up this repo a bit. :)

**Comment by zhigang1992** [comment]
Closing this one to cleanup the issue list,

Feel free to reopen this if needed or refer to https://github.com/SwiftyJSON/SwiftyJSON/issues/647
